Swap out galaxy.util.odict for core python collections.OrderedDicts

dannon opened the pull request:
Testing the waters (i.e. the full test suite) in swapping our odict for real OrderedDicts

WIP, fixes https://github.com/galaxyproject/galaxy/issues/8093, with an explanation why at https://github.com/galaxyproject/galaxy/issues/8093#issuecomment-499328125

dannon committed:
Testing the waters (i.e. the full suite) swapping our odict for real OrderedDicts

dannon committed:
Swap all odicts for OrderedDicts, remove galaxy.util.odict

dannon committed:
Attempt to fix a TODO in cwl/representation since OrderedDict is json serializable.

dannon commented:
Well, there's an optimistic refactoring first pass, time to look at a few of these more closely.

dannon committed:
Namespace conflict; force absolute import.

nsoranzo commented on lib/galaxy/tools/parameters/__init__.py:
```diff
@@ -29,7 +29,7 @@ def visit_input_values(inputs, input_values, callback, name_prefix='', label_pre
     >>> from xml.etree.ElementTree import XML
     >>> from galaxy.util.bunch import Bunch
-    >>> from galaxy.util.odict import odict
+    >>> from collections import OrderedDict
```
We don't check for import order in doctests, but for consistency move this to first?

nsoranzo commented on lib/galaxy/tools/parameters/__init__.py:
```diff
@@ -270,7 +270,7 @@ def populate_state(request_context, inputs, incoming, state, errors={}, prefix='
     >>> from xml.etree.ElementTree import XML
     >>> from galaxy.util.bunch import Bunch
-    >>> from galaxy.util.odict import odict
+    >>> from collections import OrderedDict
```
Move to first import.

nsoranzo commented on lib/galaxy/tools/toolbox/panel.py:
```diff
@@ -112,7 +112,7 @@ def to_dict(self, **kwds):
 
 
-class ToolPanelElements(odict, HasPanelItems):
+class ToolPanelElements(OrderedDict, HasPanelItems):
```
This is the only class using a method that exists only for `odict` but not for `OrderedDict`, i.e. `insert()` which in used in `update_or_append()`, `replace_tool` and `insert_tool`.

nsoranzo replied:
We discussed this on Gitter, these were the main points:
- The `insert()` method was added to `odict` in commit https://github.com/galaxyproject/galaxy/commit/7853e71dad4cfbde60a847197586769b91d2def3 .
- There is no alternative implementation that solves the tool panel use case and is as efficient as `odict`. Especially given that Python 3.6 has `OrderedDict` implemented in C, so we cannot add an `insert` method to it any more.
- If we end up keeping `odict` for this use case, it may make sense to leave it where it is since it has been distributed as part of `galaxy-lib` and now `galaxy-util` packages.
- We can remove the unused `reverse()` method of `odict`.

nsoranzo commented on lib/galaxy/util/topsort.py:
```diff
@@ -36,7 +36,7 @@
 a good deal more code than topsort itself!
 """
-from galaxy.util.odict import odict as OrderedDict
+from collections import OrderedDict as OrderedDict
```
```suggestion
from collections import OrderedDict
```

dannon replied:
Batched this with the other fixes, thanks!

nsoranzo commented on lib/galaxy/webapps/galaxy/api/users.py:
```diff
@@ -689,7 +689,7 @@ def _add_filter_inputs(self, factory, filter_types, inputs, errors, filter_type,
-        return odict([('toolbox_tool_filters', {'title': 'Tools', 'config': trans.app.config.user_tool_filters}),
+        return OrderedDict([('toolbox_tool_filters', {'title': 'Tools', 'config': trans.app.config.user_tool_filters}),
                       ('toolbox_section_filters', {'title': 'Sections', 'config': trans.app.config.user_tool_section_filters}),
                       ('toolbox_label_filters', {'title': 'Labels', 'config': trans.app.config.user_tool_label_filters})])
```
Indentation

nsoranzo commented on lib/galaxy/workflow/modules.py:
```diff
@@ -681,7 +681,7 @@ def get_inputs(self):
                 input_parameter_type.static_options[i] = tuple(option)
-        return odict([("parameter_type", input_parameter_type),
+        return OrderedDict([("parameter_type", input_parameter_type),
                       ("optional", BooleanToolParameter(None, Element("param", name="optional", label="Optional", type="boolean", value=optional)))])
```
Indentation

nsoranzo commented on test/unit/tools/test_execution.py:
```diff
@@ -185,7 +185,7 @@ def execute(self, tool, trans, **kwds):
                 return None, "Test Error Message"
 
-        return galaxy.model.Job(), odict(dict(out1="1"))
+        return galaxy.model.Job(), OrderedDict(dict(out1="1"))
```
```suggestion
        return galaxy.model.Job(), OrderedDict(out1="1")
```

dannon replied:
Batched with d83ceb4

nsoranzo committed:
Fix ``items()`` and ``keys()`` not supporting indexing on Py3

dannon committed:
Add suggestions/fixes from @nsoranzo -- thanks!

dannon committed:
Retain odict module for use in tool panel and metadata extensions, note
that it likely should not be used *unless* insert functionality is
specifically required.

dannon committed:
Swap another non-insert-necessary odict usage.

nsoranzo commented:
@dannon Why is `odict` needed for `MetadataSpecCollection` ? It doesn't use `insert()` I think, so I may be missing something.

dannon commented:
@nsoranzo I swapped it back because it was the other place where we rely on extending odict, and the swap didn't seem like it was going to be a straightforward one (https://gist.github.com/76c7890eecdd69b8bf5af286821de526 doesn't work, and I didn't want/care to dig deeper on this one right now -- bleh, that was the wrong gist, but it gets the point across; in another stash I'd correctly called super/__init__, but I don't have the traceback in front of me).

If you want to add a change that addresses that one, that'd be great.

dannon committed:
Fix symlink busted in rebase.

nsoranzo committed:
Reimplement `MetadataSpecCollection` as subclass of `OrderedDict`

nsoranzo commented:
> If you want to add a change that addresses that one, that'd be great.

Done! I've also rebased to include the integration tests fix.
